Source/Core/VideoBackends/D3D12/D3DCommandListManager.cpp

```diff
@@ -0,0 +1,352 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <queue>
+#include <vector>
+
+#include "VideoBackends/D3D12/D3DBase.h"
+#include "VideoBackends/D3D12/D3DCommandListManager.h"
+#include "VideoBackends/D3D12/D3DQueuedCommandList.h"
+#include "VideoBackends/D3D12/D3DState.h"
+#include "VideoBackends/D3D12/D3DTexture.h"
+
+#include "VideoBackends/D3D12/Render.h"
+#include "VideoBackends/D3D12/ShaderConstantsManager.h"
+#include "VideoBackends/D3D12/VertexManager.h"
+
+static const unsigned int COMMAND_ALLOCATORS_PER_LIST = 2;
+
+namespace DX12
+{
+extern StateCache gx_state_cache;
+
+D3DCommandListManager::D3DCommandListManager(
+	D3D12_COMMAND_LIST_TYPE command_list_type,
+	ID3D12Device* device,
+	ID3D12CommandQueue* command_queue,
+	ID3D12DescriptorHeap** gpu_descriptor_heaps,
+	unsigned int gpu_descriptor_heaps_count,
+	ID3D12RootSignature* default_root_signature
+	) :
+	m_device(device),
+	m_command_queue(command_queue),
+	m_gpu_descriptor_heaps_count(gpu_descriptor_heaps_count),
+	m_default_root_signature(default_root_signature)
+{
+	// Create two lists, with two command allocators each. This corresponds to up to two frames in flight at once.
+	m_current_command_allocator = 0;
+	m_current_command_allocator_list = 0;
+	for (UINT i = 0; i < COMMAND_ALLOCATORS_PER_LIST; i++)
+	{
+		for (UINT j = 0; j < ARRAYSIZE(m_command_allocator_lists); j++)
+		{
+			ID3D12CommandAllocator* command_allocator = nullptr;
+
+			CheckHR(m_device->CreateCommandAllocator(D3D12_COMMAND_LIST_TYPE_DIRECT, IID_PPV_ARGS(&command_allocator)));
+			m_command_allocator_lists[j].push_back(command_allocator);
+		}
+	}
+
+	// Create backing command list.
+	CheckHR(m_device->CreateCommandList(0, D3D12_COMMAND_LIST_TYPE_DIRECT, m_command_allocator_lists[m_current_command_allocator_list][0], nullptr, IID_PPV_ARGS(&m_backing_command_list)));
+
+#ifdef USE_D3D12_QUEUED_COMMAND_LISTS
+	m_queued_command_list = new ID3D12QueuedCommandList(m_backing_command_list, m_command_queue);
+#endif
+
+	// Copy list of default descriptor heaps.
+	for (unsigned int i = 0; i < gpu_descriptor_heaps_count; i++)
```

Comment:
Probably nitpicking a bit here, but a copy of a global, which comes from D3DDescriptorHeapManager, is there a cleaner way of doing this? Could this be replaced with a pointer to the heap manager class?